### ISSUE TITLE
[Serve] HTTPProxyActor takes too much memory

### ISSUE THREAD

**Issue by TracebaK** [ray-project/ray#37396]
### What happened + What you expected to happen

Hi, I'm running a performance benchmark on my model deployment to test the throughput of the model. The way I tested it is to post a lot of async http requests at the same time and let the serve controller to handle them. I set the number of serve replica to be exactly 1, which means I don't want any autoscaling in the test. What I found is that when the number of requests is <120 under the test setting, it worked as expected and the HTTPProxyActor only takes around 70MB memory. However, when I increase the number of requests, the memory usage of the HTTPProxyActor process increased dramatically. To be more specific, when I post 200 requests to the deployment, the memory usage of HTTPProxyActor is around 2GB. The request also takes much much longer to respond. After the computation completed, the memory used by the HTTPProxyActor is not released and I have to shutdown the serve. 

Do you have any idea why the HTTPProxyActor is taking up that much memory from the serve architecture perspective and whether it is a bug? If it's not a bug, then what can I do if I have too many incoming requests at the same time? Thank you!

### Versions / Dependencies

python == 3.7.16
ray == 2.5.1

OS: Windows 11

### Reproduction script

I'm using aiohttp and asyncio to perform the test and I think the problem should be easy to reproduce with any model. I'm using a language detection model from huggingface. 

```
import aiohttp
import asyncio
import json
from pathlib import Path
import csv
import time
from tqdm import tqdm

from zoo.config import data_dir
from zoo.api import run     # my API to depoly the serve app, it calls serve.run()


batch_size = 1
num_requests = 200


async def make_request(url, text):
    async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=None)) as session:
        async with session.post(url, json=json.dumps(text)) as response:
            return await response.text()

async def main():
    url = "http://localhost:8000/lang-det"
    with open(Path(data_dir) / 'papluca-language-identification/test.csv', newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile, delimiter=',', quotechar='"')
        next(reader)    # skip column name line
        count = 0
        tasks = []
        start = time.time()
        batch = []
        for _, row in enumerate(tqdm(reader)):
            if count > num_requests:
                break
            if len(row[1]) > 512:
                continue
            batch.append(row[1])
            if len(batch) == batch_size:
                count += 1
                tasks.append(asyncio.create_task(make_request(url, batch)))
                batch = []
            
        await asyncio.gather(*tasks)
        end = time.time()
        T = end - start
        print(f"Time elapsed: {T} s, average time per request: {T / count}")
        return 


if __name__ == '__main__':
    handle = run('LanguageDetection', 'HUGGINGFACE', 'roberta-language-detection', 
                 route_prefix='/lang-det', 
                 name='lang-det')
    asyncio.run(main())
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by TracebaK** [comment]
increase the value of `max_concurrent_queries` seems to work. But I'm still wondering what would happend when the http proxy found that no available replicas and the request stays in the queue and why it starts to consume extra memory.

**Comment by shrekris-anyscale** [comment]
Requests get queued on the HTTP Proxy until they can be processed by replicas. Increasing `max_concurrent_queries` lets your replica process more queries at once, which can help shrink this queue. That's expected behavior.

> To be more specific, when I post 200 requests to the deployment, the memory usage of HTTPProxyActor is around 2GB.

How large is each request?

> After the computation completed, the memory used by the HTTPProxyActor is not released and I have to shutdown the serve.

This sounds like a bug. The memory should get released.

Could you retry your experiment on the latest Ray nightly and see if it still happens? We've introduced some improvements to our HTTP routing that may mitigate this issue.

**Comment by TracebaK** [comment]
Hi @shrekris-anyscale ,

Thanks for your reply! For your first question, each request contains a json serialized python list of strings and the length of the list is 1. So I believe the size of the request is quite small. To reproduce the problem, you can try the following script, it no longer rely on other code from my repo. I also post the data and the model I used in the testing script here, hope it helps!

model: [papluca-language-detection](https://huggingface.co/papluca/xlm-roberta-base-language-detection)
data: [language-identification](https://huggingface.co/datasets/papluca/language-identification#additional-information)

```
import aiohttp
import asyncio
import json
from pathlib import Path
import csv
import time

from transformers import pipeline
from fastapi import FastAPI, Request
from ray import serve

from zoo.config import data_dir, model_dir


batch_size = 1
num_requests = 1000


async def make_request(url, text):
    async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=None)) as session:
        async with session.post(url, json=json.dumps(text)) as response:
            return await response.text()

async def main():
    url = "http://localhost:8000/test"
    with open(Path(data_dir) / 'papluca-language-identification/test.csv', newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile, delimiter=',', quotechar='"')
        next(reader)    # skip column name line
        count = 0
        tasks = []
        start = time.time()
        batch = []
        for _, row in enumerate(reader):
            if count > num_requests:
                break
            if len(row[1]) > 512:
                continue
            batch.append(row[1])
            if len(batch) == batch_size:
                count += 1
                tasks.append(asyncio.create_task(make_request(url, batch)))
                batch = []
            
        await asyncio.gather(*tasks)
        end = time.time()
        T = end - start
        print(tasks[0].result())
        print(f"Time elapsed: {T} s, average time per request: {T / count}")
        return 


app = FastAPI()

@serve.deployment(autoscaling_config={"min_replicas": 1,
                                      "initial_replicas": 1,
                                      "max_replicas": 1,
                                      "target_num_ongoing_requests_per_replica": 5,
                                      "upscale_delay_s": 10,
                                      "downscale_delay_s": 10}, 
                ray_actor_options={"num_cpus": 1.0, "num_gpus": 0.0})
@serve.ingress(app)
class MyDeployment:
    def __init__(self) -> None:
        self.model = pipeline(task='text-classification', model=Path(model_dir) / 'huggingface/roberta-language-detection/papluca-xlm-roberta-base')

    @app.post("/")
    async def infer(self, request: Request):
        data = await request.json()
        return self.model(json.loads(data))
    
    @app.get("/health")
    def health(self):
        return "ok"


if __name__ == '__main__':
    serve.run(MyDeployment.bind(), route_prefix='/test', name='test')
    asyncio.run(main())
```

**Comment by shrekris-anyscale** [comment]
Hi @TracebaK, we've done a lot of work addressing memory leaks in Serve recently. Most of the leaks– including ones in the `HTTPProxyActor`– were fixed in Ray 2.8, and there's one lingering leak that will be fixed in Ray 2.9. I'll close this issue for now, but if you can reproduce it on `master`, please feel free to re-open it.
